source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -235,13 +235,20 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
     decoding_state_.setCompleteBodyAvailable(true);
   }
 
-  if (!decoding_state_.sendHeaders()) {
-    ENVOY_LOG(trace, "decodeHeaders: Skipped");
-    return FilterHeadersStatus::Continue;
+  FilterHeadersStatus status = FilterHeadersStatus::Continue;
+  if (decoding_state_.sendHeaders()) {
+    status = onHeaders(decoding_state_, headers, end_stream);
+    ENVOY_LOG(trace, "onHeaders returning {}", static_cast<int>(status));
   }
 
-  const auto status = onHeaders(decoding_state_, headers, end_stream);
-  ENVOY_LOG(trace, "decodeHeaders returning {}", static_cast<int>(status));
+  // Always remove the content length in STREAMED body processing mode. In this mode, ext_proc
+  // filter can not guarantee to set the content length correctly if request body is mutated by
+  // external processor later.
+  // In http1 codec, removing content length will let it enable chunked encoding whenever
+  // feasible.
+  if (decoding_state_.bodyMode() == ProcessingMode::STREAMED) {
```

Comment:
This should only be done if body mutation is enabled.

Reply:
I feel we should do this for all body mode, i.e, changing the check to be: 
if (decoding_state_.bodyMode() != ProcessingMode::NONE) {

Leaving this change specific to STREAMED mode  leaving a logic hole when body is BUFFERED or BUFFERED_PARTIAL mode and HEADER mode is NONE.

I think our design consideration should be below to make it simple and consistent:
 As long as body mode is not NONE, then remove content-length in the headers to inform codec do chunk encoding. 

We do not have to do it one way for STREAMED mode, and do it the other way for BUFFERED/BUFFERD_PARTIAL mode. That might cause some logic holes here and there and create unnecessary complexity.

Reply:
You  can see  the discussion (I mentioned in the PR ) https://github.com/envoyproxy/envoy/issues/28515#issuecomment-1710160552) why it is only limited in STREAMED mode. i do think chunked encoding is optional for other modes and we should not break other users.

Reply:
I am wondering how do you make this change work if header mode is NONE and  body mode is BUFFERED.

In this case,  the header is already sent with the original content-length. And if the body is mutated, there is no chance for Envoy to fix the incorrect content-length, and send as chunk-encodidng.

Reply:
I think that is the question for BUFFERED mode and my change here only fixes stream mode.

I feel `header mode is NONE and body mode is BUFFERED` is not the right customer setting to mutate the body with different content length, because customer explicitly disable content_length mutation by themselves via this setting.  Maybe we can make this explicit in API/ user guide.

I imagine in the end of world.  buffer mode has two options (1) with correct configuration to update the content_length correctly, which is some OSS customers are doing  (2) remove content length to enforce chunk encoding.  

But the STREAM mode is what my PR is fixing. I am trying to avoid involving too many things in a single PR.

Reply:
The main criteria is preserving backward compatibility where it is possible to do so. In all modes where sending content-length can result in desync with modified body size, it should be removed.

Reply:
In buffered mode with correct configuration, sending content-length will NOT result in desync with modified body size because customer can set the content length correctly. This is also what OSS folks is asking for https://github.com/envoyproxy/envoy/issues/28515#issuecomment-1710160552.

I think we can spend time discussing/convincing them, but i would like to fix our issue first. The reason I think this CL alone make sense is because:
1) This CL fixes this issue and doesn't introduce any holes. It address our immediate concern.
2) This CL setup the example of how to fix potential issues in other modes, which could be done in the separate CL
3) No throw away work in this CL because 2) above.

Reply:
I agree @yanavlasov that in all modes if sending content-length will result desync with modified body size, it should be removed.  This can be easily achieved with this PR just changing the check from mode==STREAMED into mode != NONE. We are soon turning this filter into stable. Thus our solution should be complete.
Reading through the https://github.com/envoyproxy/envoy/issues/28515, please note that its Envoy define a reasonable working API and ext_proc server follows it, not the other way around. 
BTW, for supporting BUFFERED mode and BUFFERED_PARTIAL mode, I don't think there is any other viable solution other than removing the content-length in encodeHeader() and decodeHeader() considering the HEADER and body mode configuration are independant.

Reply:
This topic is outside of the scope of this PR and is what I want to avoid in the first place.  Can you discuss with @eitansuez to convince him (or potentially OSS users out there)? From his reply, he can not update their logic.

I feel whole discussion here will only delay our fix , what is the downside of submitting this PR as it is and follow up with a change if needed especially it breaks others).

I think this CL alone make sense is because:
1) This CL fixes this issue and doesn't introduce holes.
2) This CL sets up the example for follow-up PRs if needed.
3) No throw away work/tech debt in this PR because 2) above.
4) This PR is delivered in spirit of small CLs and address our concern.

Reply:
After discussing with @yanjunxiang-google offline, I am onboard with removing content length completely in all modes when body mutation is enabled. Chunked encoding is well defined in RFC spec that the server/client should support.

My intention was focusing on fixing our immediate concern while setting up the framework for addressing future issues if needed. But on second thought, The effort to include changes for all modes is small based on the changes in this PR. 

Therefore, I will update the PR with changes and clean-up to all modes. (I could wait for the response from that OSS issue though)

Reply:
Let's stick to streamed please, avoid scope creep, thanks.